File: 프로그래머스/lv2/괄호 회전하기.py
from collections import deque

def solution(s):
    s=deque(s)
    stack=[]
    cnt=0
    for _ in range(len(s)):
        s.append(s.popleft())
        flag=True
        for i in s:
            if i=='{':
                stack.append('{')
            elif i=='(':
                stack.append('(')
            elif i=='[':
                stack.append('[')
            elif i=='}':
                if not stack or stack[-1]!='{':
                    flag=False
                    break
                stack.pop()
            elif i==']':
                if not stack or stack[-1]!='[':
                    flag=False
                    break
                stack.pop()
            elif i==')':
                if not stack or stack[-1]!='(':
                    flag=False
                    break
                stack.pop()
        if stack:
            break
        if flag:
            cnt+=1


    return cnt

print(solution("}]()[{"))

File: 프로그래머스/lv2/괄호회전하기.java
import java.util.Stack;

public class 괄호회전하기 {
    public int solution(String s) {
        int answer = 0;
        for (int i=0;i<s.length();i++){
            Stack<Character> stack = new Stack<>();
            String str=s.substring(i,s.length())+s.substring(0,i);
            for (char c: str.toCharArray()){
                if (stack.isEmpty()) stack.push(c);
                else if (c==')' && stack.peek()=='(') stack.pop();
                else if (c==']' && stack.peek()=='[') stack.pop();
                else if (c=='}' && stack.peek()=='{') stack.pop();
                else stack.push(c);
            }
            if (stack.isEmpty()){
                answer++;
            }
        }

        return answer;
    }

    public static void main(String[] args) {
        괄호회전하기 ss = new 괄호회전하기();
        System.out.println(ss.solution("[](){}"));
    }
}
